exchanges/coinbene/coinbene_websocket.go

```diff
@@ -78,264 +82,296 @@ func (c *Coinbene) GenerateAuthSubs() {
+				Pair:       cp,
+				UpdateID:   orderBook.Data[0].Version,
+				UpdateTime: orderBook.Data[0].Timestamp,
+			}
+			err = c.Websocket.Orderbook.Update(&newOB)
+			if err != nil {
+				return err
+			}
+			c.Websocket.DataHandler <- wshandler.WebsocketOrderbookUpdate{Pair: newOB.Pair,
+				Asset:    asset.PerpetualSwap,
+				Exchange: c.Name,
+			}
+		}
+	case strings.Contains(result[topic].(string), "kline"):
+		var kline WsKline
+		var tempFloat float64
+		var tempKline []float64
+		err = json.Unmarshal(respRaw, &kline)
+		if err != nil {
+			return err
+		}
+		for x := 2; x < len(kline.Data[0]); x++ {
+			tempFloat, err = strconv.ParseFloat(kline.Data[0][x].(string), 64)
+			if err != nil {
+				return err
+			}
+			tempKline = append(tempKline, tempFloat)
+		}
+		p := currency.NewPairFromFormattedPairs(kline.Data[0][0].(string),
+			c.GetEnabledPairs(asset.PerpetualSwap),
+			c.GetPairFormat(asset.PerpetualSwap, true))
+		if tempKline == nil && len(tempKline) < 5 {
+			return errors.New(c.Name + " - received bad data ")
+		}
+		c.Websocket.DataHandler <- wshandler.KlineData{
+			Timestamp:  time.Unix(int64(kline.Data[0][1].(float64)), 0),
+			Pair:       p,
+			AssetType:  asset.PerpetualSwap,
+			Exchange:   c.Name,
+			OpenPrice:  tempKline[0],
+			ClosePrice: tempKline[1],
+			HighPrice:  tempKline[2],
+			LowPrice:   tempKline[3],
+			Volume:     tempKline[4],
+		}
+	case strings.Contains(result[topic].(string), "user.account"):
+		var userInfo WsUserInfo
+		err = json.Unmarshal(respRaw, &userInfo)
+		if err != nil {
+			return err
+		}
+		c.Websocket.DataHandler <- userInfo
+	case strings.Contains(result[topic].(string), "user.position"):
+		var position WsPosition
+		err = json.Unmarshal(respRaw, &position)
+		if err != nil {
+			return err
+		}
+		c.Websocket.DataHandler <- position
+	case strings.Contains(result[topic].(string), "user.order"):
+		var orders WsUserOrders
+		err = json.Unmarshal(respRaw, &orders)
+		if err != nil {
+			return err
+		}
+		for i := range orders.Data {
+			oType, err := order.StringToOrderType(orders.Data[i].OrderType)
+			if err != nil {
+				c.Websocket.DataHandler <- order.ClassificationError{
+					Exchange: c.Name,
+					OrderID:  orders.Data[i].OrderID,
+					Err:      err,
 				}
-				c.Websocket.DataHandler <- position
-			case strings.Contains(result[topic].(string), "user.order"):
-				var orders WsUserOrders
-				err = json.Unmarshal(stream.Raw, &orders)
-				if err != nil {
-					c.Websocket.DataHandler <- err
-					continue
+			}
+			oStatus, err := order.StringToOrderStatus(orders.Data[i].Status)
+			if err != nil {
+				c.Websocket.DataHandler <- order.ClassificationError{
+					Exchange: c.Name,
+					OrderID:  orders.Data[i].OrderID,
+					Err:      err,
 				}
-				c.Websocket.DataHandler <- orders
-			default:
-				c.Websocket.DataHandler <- fmt.Errorf("%s - unhandled response '%s'", c.Name, stream.Raw)
+			}
+			p := currency.NewPairFromString(orders.Data[i].Symbol)
```

Comment:
Should I use this across all websocket instances instead of `NewPairFromString`?

Reply:
For this exchange, yep. Others are context dependent, based on if the exchange doesn't send us pairs with delimiters AND if pairs are > 6 characters in length (btcusd would be fine, but dashusdt wouldn't (would become das-husdt)

Reply:
I went through websocket orders recently and tried to remove any hard coded asset types. Sadly doing this will reintroduce asset types. I'll try and work out a way of achieving future proofing and currency formatting